src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
   };
 };
 
-module PointwiseAddDistributionsWeighted = {
-  type t = pointwiseAdd;
+module DistTree = {
+  type nodeResult = [
+    | `Simple(dist)
+    // RenderedShape: continuous xyShape, discrete xyShape, total value.
+    | `RenderedShape(DistTypes.continuousShape, DistTypes.discreteShape, integral)
+  ];
 
-  let normalizeWeights = (dists: t) => {
-    let total = dists |> E.A.fmap(snd) |> E.A.Floats.sum;
-    dists |> E.A.fmap(((a, b)) => (a, b /. total));
+  let evaluateDistribution = (d: dist): nodeResult => {
+    `Simple(d)
   };
 
-  let pdf = (x: float, dists: t) =>
-    dists
-    |> E.A.fmap(((e, w)) => GenericSimple.pdf(x, e) *. w)
-    |> E.A.Floats.sum;
-
-  let min = (dists: t) =>
-    dists |> E.A.fmap(d => d |> fst |> GenericSimple.min) |> E.A.min;
-
-  let max = (dists: t) =>
-    dists |> E.A.fmap(d => d |> fst |> GenericSimple.max) |> E.A.max;
-
-  let discreteShape = (dists: t, sampleCount: int) => {
-    let discrete =
-      dists
-      |> E.A.fmap(((r, e)) =>
-           r
-           |> (
-             fun
-             | `Float(r) => Some((r, e))
-             | _ => None
-           )
-         )
-      |> E.A.O.concatSomes
-      |> E.A.fmap(((x, y)) =>
-           ({xs: [|x|], ys: [|y|]}: DistTypes.xyShape)
-         )
-      |> Distributions.Discrete.reduce((+.));
-    discrete;
+  // This is a performance bottleneck!
+  // Using raw JS here so we can use native for loops and access array elements
+  // directly, without option checks.
+  let jsContinuousCombinationConvolve: (array(float), array(float), array(float), array(float), float => float => float) => array(array((float, float))) = [%bs.raw
+  {|
+  function (s1xs, s1ys, s2xs, s2ys, func) {
```

Comment:
I didn't actually realize you could just pass in a reasonML function into js, that's nice it works so easily.